gobblin-service/src/main/java/org/apache/gobblin/service/modules/flowgraph/BaseFlowGraph.java

```diff
@@ -74,25 +90,30 @@ public synchronized boolean addDataNode(DataNode node) {
    * @return true if addition of {@FlowEdge} is successful.
    */
   @Override
-  public synchronized boolean addFlowEdge(FlowEdge edge) {
-    String srcNode = edge.getSrc();
-    String dstNode = edge.getDest();
-    if(!dataNodeMap.containsKey(srcNode) || !dataNodeMap.containsKey(dstNode)) {
-      return false;
-    }
-    DataNode dataNode = getNode(srcNode);
-    if(dataNode != null) {
-      Set<FlowEdge> adjacentEdges = this.nodesToEdges.get(dataNode);
-      if(!adjacentEdges.add(edge)) {
-        adjacentEdges.remove(edge);
-        adjacentEdges.add(edge);
+  public boolean addFlowEdge(FlowEdge edge) {
+    try {
+      rwLock.writeLock().lock();
+      String srcNode = edge.getSrc();
+      String dstNode = edge.getDest();
+      if (!dataNodeMap.containsKey(srcNode) || !dataNodeMap.containsKey(dstNode)) {
+        return false;
       }
-      this.nodesToEdges.put(dataNode, adjacentEdges);
-      String edgeId = edge.getId();
-      this.flowEdgeMap.put(edgeId, edge);
-      return true;
-    } else {
-      return false;
+      DataNode dataNode = getNode(srcNode);
+      if (dataNode != null) {
```

Comment:
Nitpick: I would the following first
```
if (dataNode == null) {
 return false
}
```

This makes the codes more stream-lined, especially in situation where we have so many bracket.

Reply:
Fixed it. Thanks!